Kernel/Storage/SD/SDHostController.cpp

```diff
@@ -0,0 +1,667 @@
+/*
+ * Copyright (c) 2023, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Format.h>
+#include <Kernel/Storage/SD/Commands.h>
+#include <Kernel/Storage/SD/SDHostController.h>
+#include <Kernel/Storage/StorageManagement.h>
+#include <Kernel/Time/TimeManagement.h>
+#if ARCH(AARCH64)
+#    include <Kernel/Arch/aarch64/RPi/SDHostController.h>
+#endif
+
+namespace Kernel {
+
+// References:
+// - BCM2835: BCM2835 ARM Peripherals: Addresses of the SD Host Controller registers
+// - SDHCI: SDHCI Simplified Host Controller Specification Version 3.0
+// - PLSS: Physical Layer Simplified Specification Version 9.00
+
+static void delay(i64 nanoseconds)
+{
+    auto start = TimeManagement::the().monotonic_time().to_nanoseconds();
+    auto end = start + nanoseconds;
+    while (TimeManagement::the().monotonic_time().to_nanoseconds() < end)
+        ;
+}
+
+constexpr u32 max_supported_sdsc_frequency = 25000000;
+
+// In "m_registers->host_configuration_1"
+// In sub-register "Clock Control"
+constexpr u32 internal_clock_enable = 1 << 0;
+constexpr u32 internal_clock_stable = 1 << 1;
+constexpr u32 sd_clock_enable = 1 << 2;
+
+// In sub-register "Software Reset"
+constexpr u32 software_reset_for_all = 0x01000000;
+
+// In Interrupt Status Register
+const u32 command_complete = 1 << 0;
+const u32 transfer_complete = 1 << 1;
+const u32 buffer_write_ready = 1 << 4;
+const u32 buffer_read_ready = 1 << 5;
+
+// PLSS 5.1: all voltage windows
+constexpr u32 acmd41_voltage = 0x00ff8000;
+// PLSS 4.2.3.1: All voltage windows, XPC = 1, SDHC = 1
+constexpr u32 acmd41_arg = 0x50ff8000;
+
+constexpr u32 block_len = 512;
+
+SDHostController::SDHostController(u32 hardware_relative_controller_id)
+    : StorageController(hardware_relative_controller_id)
+{
+}
+
+ErrorOr<void> SDHostController::reset() { return ENOTIMPL; }
+
+ErrorOr<void> SDHostController::shutdown() { return ENOTIMPL; }
+
+void SDHostController::complete_current_request(AsyncDeviceRequest::RequestResult)
+{
+    VERIFY_NOT_REACHED();
+}
+
+ErrorOr<NonnullRefPtr<SDHostController>> SDHostController::try_initialize()
+{
+#if ARCH(AARCH64)
+    auto hardware_relative_controller_id = StorageManagement::generate_relative_sd_controller_id({});
+    auto controller = TRY(adopt_nonnull_ref_or_enomem(new RPi::SDHostController(hardware_relative_controller_id)));
```

Comment:
I think this should use `new (nothrow) T`, to actually allow new to return a nullptr